File: stack_queue/239.滑动窗口最大值.go
/*
 * @lc app=leetcode.cn id=239 lang=golang
 *
 * [239] 滑动窗口最大值
 */
// @lc code=start
type MyQueue struct {
	queue []int
}

func NewMyQueue() *MyQueue {
	return &MyQueue{
		queue: make([]int, 0),
	}
}

func (q *MyQueue) Pop(x int) {
	if len(q.queue) > 0 && q.queue[0] == x {
		q.queue = q.queue[1:]
	}
}

func (q *MyQueue) Push(x int) {
	for len(q.queue) > 0 && q.queue[len(q.queue)-1] < x {
		q.queue = q.queue[:len(q.queue)-1]
	}
	q.queue = append(q.queue, x)
}

func (q *MyQueue) Front() int {
	return q.queue[0]
}

func maxSlidingWindow(nums []int, k int) []int {
	MyQueue := NewMyQueue()
	for i := 0; i < k; i++ {
		MyQueue.Push(nums[i])
	}
	res := []int{MyQueue.Front()}
	for i := k; i < len(nums); i++ {
		// 右移
		MyQueue.Pop(nums[i-k])
		MyQueue.Push(nums[i])
		res = append(res, MyQueue.Front())
	}
	return res
}

// @lc code=end

File: stack_queue/239.滑动窗口最大值.py
#
# @lc app=leetcode.cn id=239 lang=python3
#
# [239] 滑动窗口最大值
#

# @lc code=start
from collections import deque


class MyQueue:
    def __init__(self):
        self.queue = deque()  # 这里需要使用deque实现单调队列，直接使用list会超时

    def push(self, x):
        while self.queue and self.queue[-1] < x:
            self.queue.pop()
        self.queue.append(x)

    # 每次弹出的时候，比较当前要弹出的数值是否等于队列出口元素的数值，如果相等则弹出。
    # 同时pop之前判断队列当前是否为空。
    def pop(self, x):
        if self.queue and x == self.queue[0]:
            self.queue.popleft()  # list.pop()时间复杂度为O(n),这里需要使用collections.deque()

    # 查询当前队列里的最大值 直接返回队列前端也就是front就可以了。
    def front(self):
        return self.queue[0]


class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        queue = MyQueue()
        res = []
        for i in range(k):  # 先将前k的元素放进队列
            queue.push(nums[i])
        res.append(queue.front())  # result 记录前k的元素的最大值
        for i in range(k, len(nums)):
            queue.pop(nums[i - k])  # 滑动窗口移除最前面元素
            queue.push(nums[i])  # 滑动窗口前加入最后面的元素
            res.append(queue.front())  # 记录对应的最大值
        return res


# @lc code=end
